src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -34,61 +34,97 @@ namespace NodeStore {
 
 constexpr std::uint32_t DatabaseShard::ledgersPerShardDefault;
 
-DatabaseShardImp::DatabaseShardImp(Application& app,
-    std::string const& name, Stoppable& parent, Scheduler& scheduler,
-    int readThreads, Section const& config, beast::Journal j)
+DatabaseShardImp::DatabaseShardImp(
+    Application& app,
+    std::string const& name,
+    Stoppable& parent,
+    Scheduler& scheduler,
+    int readThreads,
+    Section const& config,
+    beast::Journal j)
     : DatabaseShard(name, parent, scheduler, readThreads, config, j)
     , app_(app)
+    , ctx_(std::make_unique<nudb::context>())
     , config_(config)
     , dir_(get<std::string>(config, "path"))
     , backendName_(Manager::instance().find(
-        get<std::string>(config_, "type"))->getName())
+        get<std::string>(config_, "type", "nudb"))->getName())
     , maxDiskSpace_(get<std::uint64_t>(config, "max_size_gb") << 30)
     , ledgersPerShard_(get<std::uint32_t>(
         config, "ledgers_per_shard", ledgersPerShardDefault))
     , earliestShardIndex_(seqToShardIndex(earliestSeq()))
     , avgShardSz_(ledgersPerShard_ * (192 * 1024))
 {
-    if (ledgersPerShard_ == 0 || ledgersPerShard_ % 256 != 0)
-        Throw<std::runtime_error>(
-            "ledgers_per_shard must be a multiple of 256");
+    ctx_->start();
 }
 
 DatabaseShardImp::~DatabaseShardImp()
 {
     // Stop threads before data members are destroyed
     stopThreads();
+
+    std::lock_guard<std::mutex> lock(m_);
+    complete_.clear();
+    if (incomplete_)
+        incomplete_.reset();
+    try
```

Comment:
any reason `ctx_` is special and needs exceptions caught and but not the rest of it? i.e.  could wrap the entire destructor in a try-catch? It's a little concerning that we need to catch anything during destruction, was there an instance where you saw it throwing exceptions under normal cleanup? 

```
DatabaseShardImp::~DatabaseShardImp() try {
  // cool code goes here
} catch(std::exception const& e) {
    JLOG(j_.error()) << 
        e.what(); 
}
```

Reply:
The `context` dtor will throw if any database using the object is open. It is the caller's responsibility to close databases before destroying the `context`. This try/catch is not needed since we are closing the databases beforehand.